File: problems/basic/234/solution1.cpp
// Accepted: 211ms (23.94%), 128.80MB (20.06%)

/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
public:
    bool isPalindrome(ListNode* head) {
        vector<int> vals;
        for (ListNode* node = head; node != nullptr; node = node->next) {
            vals.push_back(node->val);
        }

        // return false if not palindrome; return true otherwise
        for (int i = 0; i < vals.size() / 2; ++i) {
            if (vals[i] != vals[vals.size()-1-i]) {
                return false;
            }
        }
        return true;
    }
};

File: problems/basic/234/solution1.ts
// Accepted: 119ms (45.53%), 85.15MB (46.34%)

/**
 * Definition for singly-linked list.
 * class ListNode {
 *     val: number
 *     next: ListNode | null
 *     constructor(val?: number, next?: ListNode | null) {
 *         this.val = (val===undefined ? 0 : val)
 *         this.next = (next===undefined ? null : next)
 *     }
 * }
 */

function isPalindrome(head: ListNode | null): boolean {
    const vals: number[] = [];
    for (let node = head; node !== null; node = node.next) {
        vals.push(node.val);
    }

    for (let i = 0; i < Math.floor(vals.length/2); ++i) {
        if (vals[i] !== vals[vals.length-i-1]) {
            return false;
        }
    }
    return true;
};
